### PR TITLE
ENH: Added support for geopackage metatata

### LINKED ISSUE #713: [proposal] Add support for metadata read/write for GeoPackage
One feature that I think would be particularly useful is reading/writing metadata in GeoPackage files. I have seen that this is supported in [gdal python tests](https://trac.osgeo.org/gdal/browser/trunk/autotest/ogr/ogr_gpkg.py#L1653). However, I don't see it in the Fiona package. If it is already there and I missed it, feel free to point it out to me.

Is this a feature that you would be interested in? If so, I would be willing to create a prototype of a potential solution we could iterate on.

### PR DESCRIPTION
**Pull request by snowman2**
Closes #713

### CODE REVIEW

**Comment by snowman2** [comment]
If you have a different exception name in mind, just let me know. I just picked one that seemed somewhat reasonable.

**Comment by snowman2** [comment]
Everything but GDAL 1.11 passes. GDAL 1.11 core dumps.

**Comment by snowman2** [comment]
@sgillies Since this does not work in `GDAL` 1 and `Fiona` 2+ drops support for `GDAL` 1, thoughts on making this a `Fiona` 2.0 feature?

**Comment by rbuffat** [comment]
@snowman2 
Would it be possible to check the gdal version before accessing the code that core dumps and raising a DriverError when gdal 1.x is detected?
E.g. similar to this https://github.com/Toblerity/Fiona/blob/maint-1.8/fiona/collection.py#L81-L89

**Comment by snowman2** [comment]
> Would it be possible to check the gdal version before accessing the code that core dumps and raising a DriverError when gdal 1.x is detected?
> E.g. similar to this https://github.com/Toblerity/Fiona/blob/maint-1.8/fiona/collection.py#L81-L89

Sounds like a good idea - definitely worth a shot. :+1:

File: fiona/ogrext.pyx [hunk 1]
```diff
@@ -1207,6 +1278,80 @@ cdef class WritingSession(Session):
         gdal_flush_cache(cogr_ds)
         log.debug("Flushed data source cache")
 
+    def set_tags(self, tags, ns=None):
```
Comment (sgillies) [inline fiona/ogrext.pyx:1281]: @snowman2 what what you think about renaming this to update_tags to align with https://rasterio.readthedocs.io/en/latest/api/rasterio._io.html?highlight=set_tags#rasterio._io.DatasetWriterBase.update_tags?

Comment (snowman2) [reply fiona/ogrext.pyx]: Sounds like a good idea to me - just pushed some changes.

File: fiona/collection.py [hunk 2]
```diff
@@ -284,12 +284,12 @@ def set_tags(self, tags, ns=None):
...
         if _GDAL_VERSION_TUPLE.major < 2:
             raise GDALVersionError(
-                "set_tags requires GDAL 2+, fiona was compiled "
+                "update_tags requires GDAL 2+, fiona was compiled "
                 "against: {}".format(_GDAL_RELEASE_NAME)
             )
         if not isinstance(self.session, WritingSession):
-            raise DataIOError("Unable to set tags as not in writing mode.")
-        return self.session.set_tags(tags, ns=ns)
+            raise DataIOError("Unable to update tags as not in writing mode.")
```
Comment (sgillies) [inline fiona/collection.py:291]: @snowman2 I feel like I'm getting dangerously close to the "while you're at it..." review anti-practice, forgive me, but can you create a new fiona.errors.UnsupportedOperation exception to use here (like rasterio's https://github.com/mapbox/rasterio/blob/master/rasterio/errors.py#L103 and like the one in Python's io module)?

Comment (snowman2) [reply fiona/collection.py]: Sounds like a logical addition. I pushed the change.

**Comment by snowman2** [comment]
Thanks @snorfalorpagus, @rbuffat, and @sgillies  for reviewing this PR!

**Pull request merged**